File: sdk/java-sdk/src/main/java/com/evernym/verity/sdk/protocols/writecreddef/v0_6/WriteCredentialDefinitionV0_6.java
package com.evernym.verity.sdk.protocols.writecreddef.v0_6;

import com.evernym.verity.sdk.exceptions.UndefinedContextException;
import com.evernym.verity.sdk.exceptions.VerityException;
import com.evernym.verity.sdk.protocols.MessageFamily;
import com.evernym.verity.sdk.utils.Context;
import com.evernym.verity.sdk.utils.Util;
import org.json.JSONObject;

import java.io.IOException;

/**
 * An interface for controlling a 0.6 WriteCredentialDefinition protocol.
 */
public interface WriteCredentialDefinitionV0_6 extends MessageFamily {
    /**
     * The qualifier for the message family. Uses Evernym's qualifier.
     */
    String QUALIFIER = Util.EVERNYM_MSG_QUALIFIER;
    /**
     * The name for the message family.
     */
    String FAMILY = "write-cred-def";
    /**
     * The version for the message family.
     */
    String VERSION = "0.6";


    /**
     * @see MessageFamily#qualifier()
     */
    default String qualifier() {return QUALIFIER;}
    /**
     * @see MessageFamily#family()
     */
    default String family() { return FAMILY;}
    /**
     * @see MessageFamily#version()
     */
    default String version() {return VERSION;}

    /**
     Name for 'write' control message
     */
    String WRITE_CRED_DEF = "write";


    /**
     * Creates a RevocationRegistryConfig object that effectively disables the revocation registry
     *
     * @return a RevocationRegistryConfig object
     */
    static RevocationRegistryConfig disabledRegistryConfig() {
        JSONObject json = new JSONObject();
        json.put("support_revocation", false);
        return new RevocationRegistryConfig(json);
    }

//    static RevocationRegistryConfig revocationRegistryConfig(String tailsFile, int totalCredentials) {
//        JSONObject json = new JSONObject();
//        json.put("support_revocation", true);
//        json.put("tails_file", tailsFile);
//        json.put("max_creds", totalCredentials);
//        return new RevocationRegistryConfig(json);
//    }

    /**
     * Directs verity-application to write the specified Credential Definition to the Ledger
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @throws IOException when the HTTP library fails to post to the agency endpoint
     * @throws VerityException when wallet operations fails or given invalid context
     */
    void write(Context context) throws IOException, VerityException;

    /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     * @throws VerityException when given invalid context
     *
     * @see #write
     */
    JSONObject writeMsg(Context context) throws UndefinedContextException;

    /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     * @throws VerityException when wallet operations fails or given invalid context
     *
     * @see #write
     */
    byte[] writeMsgPacked(Context context) throws VerityException;
}

File: sdk/dotnet-sdk/src/VeritySDK/Protocols/WriteCredDef/v0_6/WriteCredentialDefinitionV0_6.cs
using System;
using System.Json;
using VeritySDK.Protocols;
using VeritySDK.Utils;

namespace VeritySDK.Protocols.WriteCredDef
{
    /// <summary>
    /// A class for controlling a 0.6 WriteCredentialDefinition protocol.
    /// </summary>
    public class WriteCredentialDefinitionV0_6 : AbstractProtocol
    {
        #region Protocol identificator

        /// <summary>
        /// The qualifier for the message family. Uses Evernym's qualifier.
        /// </summary>
        public override string qualifier() { return Util.EVERNYM_MSG_QUALIFIER; }

        /// <summary>
        /// The name for the message family.
        /// </summary>
        public override string family() { return "write-cred-def"; }

        /// <summary>
        /// The version for the message family.
        /// </summary>
        public override string version() { return "0.6"; }

        #endregion 

        #region Constructors

        /// <summary>
        /// Constructor WriteCredentialDefinitionV0_6 object
        /// </summary>
        public WriteCredentialDefinitionV0_6() { }

        /// <summary>
        /// Constructor WriteCredentialDefinitionV0_6 object
        /// </summary>
        /// <param name="threadId">the thread id of the already started protocol</param>
        public WriteCredentialDefinitionV0_6(string threadId) : base(threadId) { }

        /// <summary>
        /// Constructor WriteCredentialDefinitionV0_6 object
        /// </summary>
        public WriteCredentialDefinitionV0_6(string name, string schemaId) : this(name, schemaId, null, null) { }

        /// <summary>
        /// Constructor WriteCredentialDefinitionV0_6 object
        /// </summary>
        public WriteCredentialDefinitionV0_6(string name, string schemaId, string tag) : this(name, schemaId, tag, null) { }

        /// <summary>
        /// Constructor WriteCredentialDefinitionV0_6 object
        /// </summary>
        public WriteCredentialDefinitionV0_6(string name, string schemaId, RevocationRegistryConfig revocation) : this(name, schemaId, null, revocation) { }

        /// <summary>
        /// Constructor WriteCredentialDefinitionV0_6 object
        /// </summary>
        public WriteCredentialDefinitionV0_6(string name, string schemaId, string tag, RevocationRegistryConfig revocation)
        {
            this.name = name;
            this.schemaId = schemaId;
            this.tag = tag;
            this.revocationConfig = revocation;
        }

        #endregion 

        /// <summary>
        /// Name for 'write' control message
        /// </summary>
        string WRITE_CRED_DEF = "write";
        string name;
        string schemaId;
        string tag;
        RevocationRegistryConfig revocationConfig;

        /// <summary>
        /// Creates a RevocationRegistryConfig object that effectively disables the revocation registry 
        /// </summary>
        /// <returns>a RevocationRegistryConfig object</returns>
        public static RevocationRegistryConfig disabledRegistryConfig()
        {
            JsonObject json = new JsonObject();
            json.Add("support_revocation", false);
            return new RevocationRegistryConfig(json);
        }

        /// <summary>
        /// Directs verity-application to write the specified Credential Definition to the Ledger 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        public void write(Context context)
        {
            send(context, writeMsg(context));
        }

        /// <summary>
        /// Creates the control message without packaging and sending it.
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the constructed message (JSON object)</returns>
        public JsonObject writeMsg(Context context)
        {
            JsonObject message = new JsonObject();
            message.Add("@type", messageType(WRITE_CRED_DEF));
            message.Add("@id", WriteCredentialDefinitionV0_6.getNewId());
            message.Add("name", this.name);
            message.Add("schemaId", this.schemaId);
            addThread(message);
            if (this.tag != null) message.Add("tag", this.tag);
            if (this.revocationConfig != null) message.Add("revocationDetails", this.revocationConfig.toJson());
            return message;
        }

        /// <summary>
        /// Creates and packages message without sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the byte array ready for transport</returns>
        public byte[] writeMsgPacked(Context context)
        {
            return packMsg(context, writeMsg(context));
        }
    }
}